### PR TITLE
Decoder rework

### PR DESCRIPTION
**Pull request by yqszxx**
Signed-off-by: Boyang Han <yqszxx@gmail.com>

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- documentation

#### API Impact
None.

#### Backend Code Generation Impact
None.

#### Desired Merge Strategy

Squash: The PR will be squashed and merged

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (3.2.x, 3.3.x, 3.4.0, 3.5.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 1]
```diff
@@ -1,10 +1,78 @@
 package chisel3.util.experimental.decoder
 
+/**
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:3]: Maybe we can add some documentation about “What is a Term?”

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 2]
```diff
@@ -1,10 +1,78 @@
...
 private class Term(val value: BigInt, val mask: BigInt = 0) {
   var prime = true
 
+  /**
+    * Rule: ? covers 0 and 1, 0 covers 0, 1 covers 1
+    *       1 doesn't cover ?, 1 doesn't cover 0
+    *       0 doesn't cover ?, 0 doesn't cover 1
+    *       Term coverage means all bits coverage
+    *
+    * Example 1:
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:16]: Plz use ‘@example’

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 3]
```diff
@@ -1,10 +1,78 @@
...
+    * this       : ?0?1
+    * this.value : 0001
+    * this.mask  : 1010
+    *
+    * x          : 10??
+    * x.value    : 1000
+    * x.mask     : 0011
+    *
+    * value ^^ x.value                                       : 1001
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:25]: Use {{{ }}}} to guard these codes.

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 4]
```diff
@@ -1,10 +1,78 @@
...
+    *
+    * value ^^ x.value                                       : different
+    * (value ^^ x.value) &~ mask                             : different and I care
+    * (value ^^ x.value) &~ mask | x.mask                    : (different and valid) or the other term don't care
+    * ((value ^^ x.value) &~ mask | x.mask) &~ mask          =
+    * ((value ^^ x.value) &~ mask &~ mask) | (x.mask &~ mask)=
+    * ((value ^^ x.value) &~ mask) | (x.mask &~ mask)        : (different and I care) or (the other term don't care but I care)
+    * (((value ^^ x.value) &~ mask | x.mask) &~ mask) == 0   : no such bits that are ((different and I care) or (the other term doesn't care but I care))
+    *                                                        : all bits are (not (different and I care) and not (the other term doesn't care but I care))
+    *                                                        : all bits are (not different or I don't care) and not (the other term doesn't care but I care))
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:53]: No need too much line break here.

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 5]
```diff
@@ -1,10 +1,78 @@
...
 
+/**
+  * @param value The literal value, with don't cares being 0
+  * @param mask  Mask bits, with don't cares being 1 and cares being 0, opposite to BitPat.mask
+  */
 private class Term(val value: BigInt, val mask: BigInt = 0) {
   var prime = true
 
+  /**
+    * Rule: ? covers 0 and 1, 0 covers 0, 1 covers 1
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:11]: “Rule to define “cover”.”

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 6]
```diff
@@ -1,10 +1,78 @@
...
+    *                                                        :                                                  not (the other term needs this bit to be fine
+    *                                                        :                                                  with either 0 or 1, but I can only tolerant one value)
+    *
+    *
+    * @param x Term to be checked with
+    * @return  Whether this term covers the other term
+    */
   def covers(x: Term): Boolean = ((value ^ x.value) &~ mask | x.mask &~ mask).signum == 0
 
+  /**
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:64]: What is intersect?

**Comment by sequencer** [comment]
Hi, @yqszxx any progress on this PR?

**Comment by yqszxx** [comment]
> Hi, @yqszxx any progress on this PR?

Sorry for the delay.

File: src/main/scala/chisel3/util/experimental/decoder/Term.scala [hunk 7]
```diff
@@ -1,27 +1,131 @@
 package chisel3.util.experimental.decoder
 
+/** Term: Represent a boolean expression which all input variables ANDed together produces '1' as the output (minterm),
```
Comment (edwardcwang) [inline src/main/scala/chisel3/util/experimental/decoder/Term.scala:3]: Not sure I followed what this meant here...

**Comment by sequencer** [comment]
Months ago, I reviewed `AddressDecoder` in RC
https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomacy/AddressDecoder.scala
It also use a heuristic algorithm to solve the address space, but that algorithm is a little primitive. 
After reviewing the `Term` documentation, I think it is just another implementation to `AddressSet`.
https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomacy/Parameters.scala#L129
I wonder after Decoder API landing to chisel, can espresso also be used to replace that algorithm? I think it might be a good improvement to bus related logic.

**Comment by yqszxx** [comment]
All work done, ready for review.

File: src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      * @param y Implicant to be checked with
+      * @return Whether two implicants intersect
+      */
+    def intersects(y: BitPat): Boolean = ((x.value ^ y.value) & x.mask & y.mask) == 0
+
+    /** Check if two implicants are the same
+      * @param y Implicant to be checked with
+      * @return Whether two implicants are the same
+      */
+    def sameAs(y: BitPat): Boolean = x.mask == y.mask && x.value == y.value
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala:29]: Did you deal with width?
How about implement the equals for `BitPat`?

File: src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      * @example this = 11?0, x = 10?0 -> similar
+      * @example this = 11??, x = 10?0 -> not similar, violated rule 1
+      * @example this = 11?1, x = 10?0 -> not similar, violated rule 2
+      * @param y Implicant to be merged with
+      * @return Merge result wrapped in [[Some]] if `x` and `y` are similar,
+      *         [[None]] otherwise
+      */
+    def mergeIfSimilar(y: BitPat): Option[BitPat] = {
+      val diff = x.value ^ y.value
+      if (x.mask == y.mask && diff.bitCount == 1) { // similar
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala:46]: first instinct: `x.mask == y.mask` is too strict which might loss some corner cases, need to be discussed later.

File: src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    val m = outputs.head.getWidth
+
+    // for all outputs
+    (0 until m).map(i => {
+      // Minterms, implicants that makes the output to be 1
+      val mint = table.filter { case (_, t) => t.mask.testBit(i) && t.value.testBit(i) }.map(_._1)
+      // Maxterms, implicants that makes the output to be 0
+      val maxt = table.filter { case (_, t) => t.mask.testBit(i) && !t.value.testBit(i) }.map(_._1)
+      // Don't cares, implicants that can produce either 0 or 1 as output
+      val dc = table.filter { case (_, t) => !t.mask.testBit(i) }.map(_._1)
```
Comment (sequencer) [inline src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala:85]: These APIs should be extracted for espresso

**Review by sequencer — COMMENTED** [review]
Good job!
Looking forward to the pla circuit generator!

File: src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala (hunk 9 above)
Comment (yqszxx) [reply src/main/scala/chisel3/util/experimental/minimizer/QMCMinimizer.scala:49]: `0` + `1` -> `?`

**Pull request merged**